### PR TITLE
simplify non-flake usage

### PR DESCRIPTION
**Pull request by zimbatm**
This change breaks backward-compatibility.

* package.nix becomes default.nix
* overlay.nix disappears. Build your own overlay if you don't use
  flakes.
* replace flake-compat with flake.lock.nix that just tries to pull the
  dependencies from the lockfile.

This brings back the convention that default.nix holds all of the
packages. The main advantage is that non-flake users can now also pass
their own versions of everything to the constructore. A secondary
advantage is that `nix repl .` works and makes it easier to debug. A
third smaller advantage is that the flake.nix is now properly lean.

Related to https://github.com/nix-community/nixpkgs-fmt/issues/257 where I
noticed that the flake is itself loading flake-compat.

### CODE REVIEW

File: default.nix [hunk 1]
```diff
@@ -1 +1,122 @@
-import ./lib/overlay.nix (_: import ./packages.nix)
+{ system ? builtins.currentSystem, flakeLock ? import ./flake.lock.nix
+, nixpkgs ? import flakeLock.nixpkgs {
+  inherit system;
+  config = { };
+  overlays = [ ];
+}, lib ? nixpkgs.lib, rust-analyzer-src ? flakeLock.rust-analyzer-src }:
+with builtins;
+with lib;
```
Comment (figsoda) [inline default.nix:8]: This should be able to replace `flake.lock.nix`, also I don't think flakeLock should be an argument here
```suggestion
with builtins;

let
  getFlake = name:
    with (fromJSON (readFile ./flake.lock)).nodes.${name}.locked;
    fetchTarball {
      url = "https://github.com/${owner}/${repo}/archive/${rev}.tar.gz";
      sha256 = narHash;
    };

in { system ? currentSystem
, nixpkgs ? import (getFlake "nixpkgs") { inherit system; }, lib ? nixpkgs.lib
, rust-analyzer-src ? getFlake "rust-analyzer-src" }:

with lib;

```

File: flake.nix [hunk 2]
```diff
@@ -10,123 +10,19 @@
...
-            '';
-          };
-        }) {
-          aarch64-darwin = "aarch64-apple-darwin";
-          aarch64-linux = "aarch64-unknown-linux-gnu";
-          i686-linux = "i686-unknown-linux-gnu";
-          x86_64-darwin = "x86_64-apple-darwin";
-          x86_64-linux = "x86_64-unknown-linux-gnu";
-        };
+    packages = builtins.mapAttrs (system: _:
```
Comment (figsoda) [inline flake.nix:13]: use [`genAttrs`](https://github.com/nixos/nixpkgs/blob/3f4ce46a4702940cb7d1d12349c0407576225605/lib/attrsets.nix#L100-L107)

**Review by figsoda — CHANGES_REQUESTED** [review]
it would be better if `default.nix` is more `callPackage` friendly

**Comment by zimbatm** [comment]
the history needs to be squashed but I wanted to keep the changes apparent for now

**Comment by figsoda** [comment]
Changing from `packages` to `legacyPackages` would basically break it for everybody. There are other ways of addressing this (e.g. making `complete` an alias to `complete.toolchain`) and failing `nix flake check` is not a big issue and can be addressed in another pr.

**Comment by zimbatm** [comment]
yeah. I'll revert that last commit because there is no way to keep the current structure and be `flake check` compatible.

**Pull request merged**